File: src/main/java/nox/AspectJ.java
/**
 * Copyright (c): Java port of https://github.com/eveoh/gradle-aspectj
 */
package nox;

import nox.compile.AjcTask;
import org.gradle.api.GradleException;
import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.Task;
import org.gradle.api.artifacts.ConfigurationContainer;
import org.gradle.api.plugins.JavaBasePlugin;
import org.gradle.api.plugins.JavaPluginConvention;
import org.gradle.api.tasks.SourceSet;
import org.gradle.api.tasks.TaskContainer;


public class AspectJ implements Plugin<Project> {

	@Override
	public void apply(Project project) {
		project.getPluginManager().apply(JavaBasePlugin.class);

		if (!project.hasProperty("aspectjVersion")) {
			throw new GradleException("You must set the property 'aspectjVersion' before applying the aspectj plugin");
		}

		ConfigurationContainer confs = project.getConfigurations();
		TaskContainer tasks = project.getTasks();

		if (confs.findByName("ajtools") == null) {
			Object aspectjVersion = project.getProperties().get("aspectjVersion");
			confs.create("ajtools");
			project.getDependencies().add("ajtools", "org.aspectj:aspectjtools:" + aspectjVersion);
			project.getDependencies().add("compile", "org.aspectj:aspectjrt:" + aspectjVersion);
		}

		JavaPluginConvention javaConv = (JavaPluginConvention) project.getConvention().getPlugins().get("java");


		for (SourceSet projectSourceSet: javaConv.getSourceSets()) {
			NamingConventions namingConventions = projectSourceSet.getName().equals("main") ? new MainNamingConventions() : new DefaultNamingConventions();
			String aspectPathConfName = namingConventions.getAspectPathConfigurationName(projectSourceSet);
			String aspectInpathConfName = namingConventions.getAspectInpathConfigurationName(projectSourceSet);
			for (String configuration : new String[]{aspectPathConfName, aspectInpathConfName}) {
				if (confs.findByName(configuration) == null) {
					confs.create(configuration);
				}
			}

			if (!projectSourceSet.getAllJava().isEmpty()) {
				String aspectTaskName = namingConventions.getAspectCompileTaskName(projectSourceSet);
				Task javaTask = tasks.getByName(namingConventions.getJavaCompileTaskName(projectSourceSet));

				AjcTask ajc = tasks.create(aspectTaskName, AjcTask.class);
				// aspectTaskArgs.put("overwrite", true);
				ajc.setDescription("Compiles AspectJ Source for " + projectSourceSet.getName() + " source set");
				ajc.sourceSet = projectSourceSet;
				ajc.getInputs().files(projectSourceSet.getAllJava());
				ajc.getOutputs().dir(projectSourceSet.getOutput().getClassesDir());
				ajc.aspectpath = confs.findByName(aspectPathConfName);
				ajc.ajInpath = confs.findByName(aspectInpathConfName);

				ajc.setDependsOn(javaTask.getDependsOn());
				ajc.dependsOn(ajc.aspectpath);
				ajc.dependsOn(ajc.ajInpath);
				javaTask.deleteAllActions();
				javaTask.dependsOn(ajc);
			}
		}
	}

	private interface NamingConventions {

		String getJavaCompileTaskName(SourceSet sourceSet);

		String getAspectCompileTaskName(SourceSet sourceSet);

		String getAspectPathConfigurationName(SourceSet sourceSet);

		String getAspectInpathConfigurationName(SourceSet sourceSet);
	}

	private static class MainNamingConventions implements NamingConventions {

		@Override
		public String getJavaCompileTaskName(final SourceSet sourceSet) {
			return "compileJava";
		}

		@Override
		public String getAspectCompileTaskName(final SourceSet sourceSet) {
			return "compileAspect";
		}

		@Override
		public String getAspectPathConfigurationName(final SourceSet sourceSet) {
			return "aspectpath";
		}

		@Override
		public String getAspectInpathConfigurationName(final SourceSet sourceSet) {
			return "ajInpath";
		}
	}

	private static class DefaultNamingConventions implements NamingConventions {

		@Override
		public String getJavaCompileTaskName(final SourceSet sourceSet) {
			return "compile" + capitalize(sourceSet.getName()) + "Java";
		}

		@Override
		public String getAspectCompileTaskName(final SourceSet sourceSet) {
			return "compile" + capitalize(sourceSet.getName()) + "Aspect";
		}

		@Override
		public String getAspectPathConfigurationName(final SourceSet sourceSet) {
			return sourceSet.getName() + "Aspectpath";
		}

		@Override
		public String getAspectInpathConfigurationName(final SourceSet sourceSet) {
			return sourceSet.getName() + "AjInpath";
		}

		private String capitalize(String name) {
			return name.substring(0, 1).toUpperCase() + name.substring(1);
		}
	}


}

File: src/main/kotlin/nox/AspectJ.kt
/**
 * Copyright (c): Java port of https://github.com/eveoh/gradle-aspectj
 */
package nox

import nox.compilation.AjcTask
import org.gradle.api.GradleException
import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.plugins.JavaBasePlugin
import org.gradle.api.plugins.JavaPluginConvention
import org.gradle.api.tasks.SourceSet


class AspectJ : Plugin<Project> {

	override fun apply(project: Project) {
		project.pluginManager.apply(JavaBasePlugin::class.java)

		if (!project.hasProperty("aspectjVersion")) {
			throw GradleException("You must set the property 'aspectjVersion' before applying the aspectj plugin")
		}

		val confs = project.configurations
		val tasks = project.tasks

		if (confs.findByName("ajtools") == null) {
			val aspectjVersion = project.properties["aspectjVersion"]
			confs.create("ajtools")
			project.dependencies.add("ajtools", "org.aspectj:aspectjtools:" + aspectjVersion)
			project.dependencies.add("compile", "org.aspectj:aspectjrt:" + aspectjVersion)
		}

		val javaConv = project.convention.plugins["java"] as JavaPluginConvention


		for (projectSourceSet in javaConv.sourceSets) {
			val namingConventions = if (projectSourceSet.name == "main") MainNamingConventions() else DefaultNamingConventions()
			val aspectPathConfName = namingConventions.getAspectPathConfigurationName(projectSourceSet)
			val aspectInpathConfName = namingConventions.getAspectInpathConfigurationName(projectSourceSet)
			for (configuration in arrayOf(aspectPathConfName, aspectInpathConfName)) {
				if (confs.findByName(configuration) == null) {
					confs.create(configuration)
				}
			}

			if (!projectSourceSet.allJava.isEmpty) {
				val aspectTaskName = namingConventions.getAspectCompileTaskName(projectSourceSet)
				val javaTask = tasks.getByName(namingConventions.getJavaCompileTaskName(projectSourceSet))
				val classesTask = tasks.getByName("classes")

				val ajc = tasks.create(aspectTaskName, AjcTask::class.java)
				// aspectTaskArgs.put("overwrite", true);
				ajc.description = "Compiles AspectJ Source for " + projectSourceSet.name + " source set"
				ajc.sourceSet = projectSourceSet
				ajc.inputs.files(projectSourceSet.allJava)
				ajc.outputs.dir(projectSourceSet.output.classesDirs.singleFile)
				ajc.aspectpath = confs.findByName(aspectPathConfName)
				ajc.ajInpath = confs.findByName(aspectInpathConfName)

				ajc.dependsOn(javaTask.taskDependencies)
				ajc.dependsOn(ajc.aspectpath)
				ajc.dependsOn(ajc.ajInpath)
				javaTask.setActions(emptyList())
				classesTask.dependsOn(ajc)
				javaTask.mustRunAfter(ajc)
			}
		}
	}

	private interface NamingConventions {

		fun getJavaCompileTaskName(sourceSet: SourceSet): String

		fun getAspectCompileTaskName(sourceSet: SourceSet): String

		fun getAspectPathConfigurationName(sourceSet: SourceSet): String

		fun getAspectInpathConfigurationName(sourceSet: SourceSet): String
	}

	private class MainNamingConventions : NamingConventions {

		override fun getJavaCompileTaskName(sourceSet: SourceSet): String {
			return "compileJava"
		}

		override fun getAspectCompileTaskName(sourceSet: SourceSet): String {
			return "compileAspect"
		}

		override fun getAspectPathConfigurationName(sourceSet: SourceSet): String {
			return "aspectpath"
		}

		override fun getAspectInpathConfigurationName(sourceSet: SourceSet): String {
			return "ajInpath"
		}
	}

	private class DefaultNamingConventions : NamingConventions {

		override fun getJavaCompileTaskName(sourceSet: SourceSet): String {
			return "compile" + capitalize(sourceSet.name) + "Java"
		}

		override fun getAspectCompileTaskName(sourceSet: SourceSet): String {
			return "compile" + capitalize(sourceSet.name) + "Aspect"
		}

		override fun getAspectPathConfigurationName(sourceSet: SourceSet): String {
			return sourceSet.name + "Aspectpath"
		}

		override fun getAspectInpathConfigurationName(sourceSet: SourceSet): String {
			return sourceSet.name + "AjInpath"
		}

		private fun capitalize(name: String): String {
			return name.substring(0, 1).toUpperCase() + name.substring(1)
		}
	}


}
